src/modules/content-s3/s3.c

```diff
@@ -45,18 +41,19 @@ static S3Status response_props_cb(const S3ResponseProperties *properties, void *
 
 static void response_complete_cb(S3Status status, const S3ErrorDetails *error, void *data)
 {
-    status_g = status;
+    struct cb_data *_data = data;
+    _data->status = status;
 }
 
 /* Writes a chunk of 'data' to s3, returning
  * the size of the data written.
  */
 static int put_object_cb(int buff_size, char *buff, void *data)
 {
-    struct put_cb_data *cb_data = data;
+    struct cb_data *_data = data;
 
-    memcpy(buff, cb_data->data + cb_data->count, buff_size);
-    cb_data->count += buff_size;
+    memcpy(buff, _data->data + _data->count, (buff_size < _data->size - _data->count ? buff_size : _data->size - _data->count));
+    _data->count += buff_size;
 
     return buff_size;
```

Comment:
If there's a chance that `buff_size` is not the amount of data coiped (e.g. if `size - count` is less), then you would want to return `size - count`.

I would probably do something like (pseudocodeish):
```c
int size = cb_data->size - cb_data->count;
if (size > buff_size)
    size = buff_size
memcpy (buff, cb_data->data + cb_data->count, size);
cb_data->count += size;
return size;
```